### ISSUE TITLE
Crash / deadlock when used in a thread pool

### ISSUE THREAD

**Issue by piotr-yuxuan** [clj-python/libpython-clj#110]
Here is a minimal working example of this bug: https://github.com/piotr-yuxuan/minimal-bug

I'm trying to use a round-trip interop: I run a Clojure program which initialises a Python runtime. Within this Python runtime, I want to call some Clojure functions, so that I can reuse robust, efficient Clojure libraries I know form within Python code. This bug appeared when using [safely](https://github.com/BrunoBonacci/safely) but is unrelated to this library and can be reproduced alone.

Several different examples are provided with log files. I have noticed that the call to a Python method fails when done from another thread, but when you force the termination of the evaluation, stdout suggests it is actually Python method is actually executed. I find it weird. Interestingly, a vanilla thread works well, but ExecutorServices don't.

It's the most puzzling thing I have seen so far when using `libpython-clj`. Otherwise the interop is well-rounded and pretty neat, well done 👍

```
$ lein run safely-clj-fail
OpenJDK 64-Bit Server VM warning: Options -Xverify:none and -noverify were deprecated in JDK 13 and will likely be removed in a future release.
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
before in thunk
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x000000013c3e25a1, pid=11408, tid=28163
#
# JRE version: OpenJDK Runtime Environment AdoptOpenJDK (14.0+36) (build 14+36)
# Java VM: OpenJDK 64-Bit Server VM AdoptOpenJDK (14+36, mixed mode, sharing, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [Python+0xab5a1]  _PyEval_EvalFrameDefault+0x55c9
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/piotr-yuxuan/src/github.com/piotr-yuxuan/minimal-bug/hs_err_pid11408.log
#
# If you would like to submit a bug report, please visit:
#   https://github.com/AdoptOpenJDK/openjdk-support/issues
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
```

**Comment by piotr-yuxuan** [comment]
Addendum: after using `jstack` on case `new-thread-hang`, it looks like it's probably a dead lock situation:
- `(py/py.. (py/import-module "MinimalBug") (call_me_baby))` acquire a lock on the interpreter
- It calls `ClojureBridge.safely`, which in turn tries invoke Clojure code and ends up evaluating `((py/as-jvm f))`.
- This latter evaluation tries to acquire the lock on the interpreter, and wait as long the evaluation of the first form isn't completed.

This explains why the termination of the hung form actually shows the execution of `call_me_baby`. Any comment / objection  / help on how to avoid that? 🙂

**Comment by jjtolton** [comment]
Hiya @piotr-yuxuan!  Yes it is a known issue.  Currently, multithreaded and multiprocessed Python code is not supported by libpython-clj.  Early discussions with the community indicated that the preference was to focus on single-threaded performance, so that's where we put the emphasis.  There are two workarounds.

1)  If you write the section of multithreaded code entirely in Python and then call that section of code from Clojure, it will probably work.  
2)  Use other standard methods of parallel development -- i.e., horizontally scale your containers and farm out work to them via proxy.  
3 (optional)) this is an open area of research with a lot of unknowns -- feel free to experiment and if you come up with anything interesting PRs are highly welcome!  Even in terms of documentation.

**Comment by piotr-yuxuan** [comment]
Thank you for your quick, thorough reply. I appreciate the community-driven approach you refer to! 👍 I have a couple of feedback:

- Regarding your first point, my goal has been to be able to use some Clojure libraries from Python (here: safely). The thing is safely does nicely work almost all the time, except when thread pool is used.

- Currently in the example code a Python global variable is redefined from Clojure-land so that Python can call Clojure code; would you know any more idiomatic way to do so? The the best of my knowledge, the doc doesn't provide such.

- Regarding your last point, perhaps we are not far from a elegant solution as clj->Python->clj-thread->Python appears to be working [from a vanilla thread](https://github.com/piotr-yuxuan/minimal-bug/blob/cdc52387082853d1bb0831df52e87fae24482902/src/clojure/core.clj#L100-L109), but not from a [thread pool](https://github.com/piotr-yuxuan/minimal-bug/blob/cdc52387082853d1bb0831df52e87fae24482902/src/clojure/core.clj#L52-L80).

I will try to write some Python methods which puts the interop logic in a thread, start it, and release the gil to see if the other thread can get it, do its stuff and finally release it, perhaps you would have some hints about that?

**Comment by jjtolton** [comment]
A few thoughts:

First of all, there is a hope of some moonshot research: https://github.com/clj-python/libpython-clj/issues/94  
I still haven't had time to investigate, but I will say WITH GREAT CAUTION AND CONSERVATISM that *potentially* this technique could be use to invert the control of the process and use Python to control Clojure (instead of Clojure controlling Python).  If that is the case, then you could potentially use Python multithreading to call Clojure code (which, might in turn call further Python code -- I have no idea if that will make the universe collapse or not)

Besides that, currently in production I do not currently attempt multithreading.  However if you'd like to post a gist or share some code I'd be happy to hack on it with you and see what we can come up with :)

**Comment by cnuernber** [comment]
I think what is happening is just the [GIL](https://realpython.com/python-gil/#:~:text=The%20Python%20Global%20Interpreter%20Lock%20or%20GIL%2C%20in%20simple%20words,at%20any%20point%20in%20time.) is being held by one thread and then python is attempting to be called on another thread.

For the truly bold, you could potentially release the GIL in your clojure code.  Then the system *should* grab it automatically when it tries to call into python.

Great caution is advised :-).  This is what numpy and various C++ libraries do; they release the GIL when they know they have control and python won't be called for a minute.  This allows other threads to call into python.

**Comment by jjtolton** [comment]
Throw caution to the wind!!  Listen to it go wooooooosh

**Comment by piotr-yuxuan** [comment]
TL;DR: YOLO mates, life is too short for caution, let's have fun instead 😉

Thanks to the two of you for these really interesting hints. Indeed I pretty much agree it is most probably a GIL issue. As suggested in the [second message](https://github.com/clj-python/libpython-clj/issues/110#issuecomment-662757184) of this thread, the deadlock can be described in generic terms as:
> Clojure code calls some Python code (which acquires GIL) which calls some other Clojure code which runs a thread which waits for the GIL.

In more palatable, less generic terms, please let me first add some more context. My goal as a Clojure-writting engineer is to enable Python-writting data scientists to write some Python code so that they can get a service up and running with minimal work on the engineering side. The overall goal is to make them as free as they want, so they don't depend on the priorities of the engineering team for their own projects. Hence, the Clojure code provides an http server with logging, monitoring, and so on. It also provides expose some Clojure-land helpers to Python code, for example [`safely`](https://github.com/BrunoBonacci/safely) as we want our services to be resilient. The [example repository](https://github.com/piotr-yuxuan/minimal-bug) linked in the opening message is a rather accurate projection of such service code. Of course it has been trimmed down to the bare minimum, but essentially it still contains the minimal code which produces the deadlock we're talking about.

(If you'd be interested in it, I would probably be able to share the final, working state of such Python-Clojure server once this bug is fixed or worked around.)

Within this context, the bug can now be described as:
> Clojure server starts and expose an API. When a request is received, it calls some Python data-science code (amd axquire the GIL). It turns out that this latter code uses `safely` to wrap some Python-land side-effect (call to an external API). `safely` is used with a circuit-breaker (one of its salient features), so another thread is created in which the actually wrapped Python side-effect will be called. It tries to acquire the GIL, which obviously it can't because it's locked by its caller. After a timeout, it throws an exception. `clj-python` doesn't like it and crash the VM.

Indeed it is control inversion.

As a simple workaround, I could write more detailed Clojure helpers so that they don't wrap any Python code.

Now it's a bit far from the initial bug, but I wonder about how this whole mechanism would behave under the pressure of some hundreds of requests per second. If I get it correctly, as is it would be limited by the GIL, hence at most 1 request at every moment, right?

**Comment by jjtolton** [comment]
Wow!  Ambitious project.  I can't speak for @cnuernber but I can tell you the assumptions I have when I'm writing contributions for libpython-clj and when I write libpython-clj production code: 

**data-first priority**

Meaning, **if** (BIG if) I were in your situation, my thoughts would be towards transmission of data back and forth between the minimal scientist rig and your resiliency rig (as opposed to semi-remote code execution).  So for instance, perhaps use Python decorators to put data, the results -- hell, even the whole damn pickled function -- on a queue over the wire.  Pick it up in a Clojure channel via some middleware.  Farm it out to whatever libpython-clj process/container (I'll use a farm of docker containers to do this sort of "fanout" computation work) is appropriate.  Execute the code on the libpython-clj container and put the result (or failure) back on the results queue and return the work syncronously (or asnycronously) to the Python user.  This is a proven pattern for heavy computational libpython-clj work.  

In other words, in my highly opinionated opinion, libpython-clj has been optimized around performant moving of **single-threaded** data operations and for exploratory development on the REPL.  So to get around that limitation, use a fanout strategy of your choice (I prefer docker containers), a queue or some equivalent mechanism, a Clojure controller (think core.async or perhaps a Ring server), and the data source (the scientist's Python code).  

I am **not** suggesting that I know how you should run your project -- I'm just trying to give you some insight into how **I write my libpython-clj code** and where I *feel* the focus has been.  I'm arming you with this understanding to give you insight into the *design intent* so that you can potentially find a way to hack it to **your** needs!!  

As you can tell, we like to stretch languages to the limit around here, so if you can find a way to stretch libpython-clj, we're all about it!

In the meantime, this weekend if I can get some time I'll see if I can offer any insight more relevant to your original vision.

**Comment by jjtolton** [comment]
here is a very, VERY crude example of what I mean:

```python 
@decorator
def funcwrap(f):
    def inner(*args):
      # q.put is hypothetical send-to-clojure on wire, get back task_id
        task_id = q.put(json.dumps({"fn": f.__name__, "args": args})) 
        while True:
            time.sleep(0.01)
            res = q.get(task_id) # potentially get result back from task_id
            if res:
                return res
    return inner
```
This is a very crude decorator but hopefully it demonstrates the point.  

To be really sophisticated (aka throwing caution to the wind), if you were using `libpython-clj`, `q.put` could be a proxy for `a/offer!` and put the JSON on a Clojure core.async channel.  That could then get farmed out to one of the waiting containers.  I haven't experimented with this, but I imagine `q.get` could a proxy for `a/<!!` or `a/<!` but that would be trickier.  

In fact, if you were using `core.async`, you wouldn't even need to use `json.dumps` since libpython-clj would handle the bridging for you.  

Alternatively `q.put` could be a more traditional "send to kafka/rabbit/redis/whatever".  

I offer this up not as a "you should do this" but simply as an illustration of how I plan out my libpython-clj work.

**Comment by cnuernber** [comment]
[Embedding Clojure in Python](https://clj-python.github.io/libpython-clj/embedded.html) allows you to use both Clojure and Python multithreading.  So you could potentially use both JVM and process-level threads.  

I am not sure this solves this issue really but it gives you I think the largest toolbox you can get to work around these types of issues.
